### PR TITLE
Server/project

### PR DESCRIPTION
**Pull request by shubham-cpp**
# Project 
- Controller
- Model
- Routes

# Tasks
- When deleting task, remove it from project

# User
- When deleting, check if user is `currentOwner` of a task or `projectOwner` of project, if so don't allow to delete the user.
- When deleting, make sure we remove the user from `project.members` array

### Additionals
- Now using authenticate middleware instead of redundant code for all requests.
- Changed the way to handle errors in express using middleware.

Future note, separate out all middlewares into their own folders like `controller/`, `models/`, etc

### CODE REVIEW

**Comment by shubham-cpp** [comment]
Hi @isandeepkadam, could you review this one

File: server/src/models/Project.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      type: String,
+      required: true,
+      minlength: 5,
+    },
+    projectId: {
+      type: String,
+      required: true,
+    },
+    projectOwner: {
+      type: Schema.Types.ObjectId,
```
Comment (isandeepkadam) [inline server/src/models/Project.ts:30]: I think project owner could also be multiple users (co-shared)? just an opinion may be, single is also fine for now.

File: server/src/routes/User.ts [hunk 2]
```diff
@@ -25,10 +26,11 @@ userRouter
...
   .post('/login', validate(loginUserValidationSchema), loginUser)
   .post('/refresh', validate(refreshTokenValidationSchema), tokenRefresh)
+  .get('/:id/projects', authenticateToken, validate(deleteUserValidationSchema), getProjectsByOwner)
   .delete('/logout', validate(refreshTokenValidationSchema), logoutUser)
-  .get('/:id', validate(deleteUserValidationSchema), getUser) // this seems wrong by naming
-  .put('/:id', validate(updateUserValidationSchema), updateUser)
-  .patch('/:id', validate(updateUserValidationSchema), updateUser)
-  .delete('/:id', validate(deleteUserValidationSchema), deleteUser)
-  .get('/:id/tasks', validate(getUserTaskValidationSchema), getUserTasks);
+  .get('/:id', authenticateToken, validate(deleteUserValidationSchema), getUser) // this seems wrong by naming
```
Comment (isandeepkadam) [inline server/src/routes/User.ts:31]: Having similar name for schema is kind of confusing like deleteUserValidationSchema
May be we can rename it to something like userValidationSchema.

File: server/src/controllers/User.ts [hunk 3]
```diff
@@ -98,59 +75,43 @@ export const createUser = (req: Request, res: Response) => {
...
+      throw new NotFound('User not found.');
     }
     if (name) user.name = name;
     if (email) user.email = email;
-    if (password) user.password = password;
+    if (password) {
+      const salt = await bcrypt.genSalt(10);
+      const hashedPassword = await bcrypt.hash(password, salt);
+      user.password = hashedPassword;
+    }
```
Comment (isandeepkadam) [inline server/src/controllers/User.ts:106]: good catch

File: server/src/models/Project.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    },
+    description: {
+      type: String,
+      required: true,
+      minlength: 5,
+    },
+    projectId: {
+      type: String,
+      required: true,
+    },
```
Comment (isandeepkadam) [inline server/src/models/Project.ts:28]: I think we String is right, while creating task schema I used number for projectID

**Comment by isandeepkadam** [comment]
@shubham-cpp review done

File: server/src/models/Project.ts (hunk 1 above)
Comment (shubham-cpp) [reply server/src/models/Project.ts:30]: The idea of projectOwner is which client owns the project currently(generally).

File: server/src/routes/User.ts (hunk 2 above)
Comment (shubham-cpp) [reply server/src/routes/User.ts:31]: correct will rename this to `userIdValidationSchema`

File: server/src/models/Project.ts (hunk 4 above)
Comment (shubham-cpp) [reply server/src/models/Project.ts:28]: Yes changing that,

**Comment by shubham-cpp** [comment]
Check now

**Review by isandeepkadam — APPROVED** [review]

**Comment by isandeepkadam** [comment]
Looks good

File: server/src/models/Project.ts (hunk 1 above)
Comment (isandeepkadam) [reply server/src/models/Project.ts:30]: we can discuss this later

**Pull request merged**